File: src/Exercise69.java
public class Exercise69 {
    class Solution {
        public int mySqrt(int x) {
            int lo = 1, hi = x;
            while (lo <= hi) {
                int mid = lo + (hi - lo) / 2;
                if (mid  > x/mid)//avoid mid*mid>maxint
                    hi = mid - 1;
                else if (mid < x/mid)
                    lo = mid + 1;
                else
                    return mid;
            }
            return hi;

        }
    }

    public static void main(String[] args) {
        Exercise69 exercise = new Exercise69();
        Solution solution = exercise.new Solution();
        System.out.println(solution.mySqrt(8));
    }

}

File: src/Math/Exercise69.cpp
#include <iostream>
using namespace std;

class Solution
{
public:
    int mySqrt(int x)
    {
        if (x <= 1)
            return x;
        int left = 0, right = x;
        while (left < right)
        {
            int mid = left + (right - left) / 2;
            if (mid > x / mid)
            {
                right = mid - 1;
            }
            else if (mid < x / mid)
            {
                left = mid + 1;
            }
            else
                return mid;
        }
        return left * left > x ? left - 1 : left;
    }
};

int main(){
    int x = 2147395599;
    Solution sl;
    cout << sl.mySqrt(x);
}